docs/SPEC-composition.md

```diff
@@ -0,0 +1,173 @@
+Flux Composition
+==
+
+The composition capabilities of the flux LSP server enable clients to create a query builder experience without deep knowledge of the flux internals. Composition will create a "Composition-owned" statement in a given flux file, and will be able to add/remove filter calls to that statement. The composition rules apply as follows:
+
+  - Composition must never generate a query that cannot be executed.
+  - Composition should never generate a query that a user with sufficient flux knowledge would never write, e.g. `filter(fn: (r) => r.foo == "bar") |> filter(fn: (r) => r.foo == "bar")`
+  - Composition should never generate a query that results in immediate diagnostic messages.
+  - In the event of conflicts or logical ambiguity, Composition should default to erroring rather than introducing complex assumption logic to attempt recovery.
+
+Given these rules, this document will not attempt to document the resulting flux code generated from each of these requests, as the resulting flux could change as requirements evolve and the corresponding flux would become outdated.
+
+Composition Initialize
+--
+
+The composition initialize request is sent from the client to the server to identify a file that will be used to compose queries. This call must be made before any other Composition requests, otherwise those requests will error. It will result in a `workspace/applyEdit` request from the server to the client with the associated changes to the file. The request can also be used to reset the composition query state by calling it again, and can be considered idempotent.
+
+*Request*
+  - method: `fluxComposition/initialize`
+  - params: `CompositionInitializeParams`
+
+```
+interface CompositionInitializeParams {
+    /*
+     * The text document.
+     */
+    text_document: TextDocumentIdentifier;
+
+    /*
+     * The bucket to initialize the composition query.
+     */
+    bucket: string;
+
+    /*
+     * The measurement to initialize the composition query.
+     */
+    measurement?: string;
+}
+```
+
+*Response*
+  - result: `null`
+  - error: code and message set in case an exception happens during the initialize request.
+
+Add Measurement Filter
+--
+
+The add measurement filter request is sent from the client to the server to add a measurement filter. If a measurement is not already specified, e.g. it was not specified in the initialization and not called previously, it will result in a `workspace/applyEdit` request from the server to the client with the associated filter addition. If a measurement filter already exists, an error will be reported in the form of a `window/showMessage` request from the server to the client with the associated information.
```

Comment:
YES!

Can we also look into having our error handlers (on the worker thread) post an `window/showMessage` back to the monaco-editor? To prompt the user to reload...as a short term solution.

Reply:
Yeah, that's the thing I want to sort out with you: if we send errors back, can we have the front end find these errors and present them to the user in certain cases?

Reply:
(Acknowledging that the general error handling case is outside the scope of this document, but still something we should definitely talk about now that @juliajames12 has some really nice UI for presenting errors to the user.)

Reply:
We are now doing something else besides the `window/showMessage`. It's work we hope to do shortly after the initial composition work. Idea is to solve the general problem -- because we now use the LSP as the source of flux injection (to ensure correctness of flux script), we still need the UI (outside of the monaco-editor client) to do :all-the-things: in response.

https://github.com/influxdata/ui/issues/5305